In [1]:
import pandas as pd
import numpy as np
import statsmodels as st
import statsmodels.api as sm
import statsmodels.formula.api as smf

In [2]:
from scipy import stats

In [3]:
df = pd.read_excel('mult_regr.xlsx')

In [3]:
df.head()

HeatFlux   East  South  North
0     271.8  33.53  40.55  16.66
1     264.0  36.50  36.19  16.46
2     238.8  34.66  37.31  17.66
3     230.7  33.13  32.52  17.50
4     251.6  35.75  33.71  16.40

In [4]:
df.shape

(29, 4)

In [7]:
flux = df[df.columns[0]]

In [9]:
flux.head()

0    271.8
1    264.0
2    238.8
3    230.7
4    251.6
Name: HeatFlux, dtype: float64

In [10]:
flux[:4]

0    271.8
1    264.0
2    238.8
3    230.7
Name: HeatFlux, dtype: float64

In [11]:
flux = df.HeatFlux

In [13]:
flux.head()

0    271.8
1    264.0
2    238.8
3    230.7
4    251.6
Name: HeatFlux, dtype: float64

In [14]:
flux[:5]

0    271.8
1    264.0
2    238.8
3    230.7
4    251.6
Name: HeatFlux, dtype: float64

In [18]:
factor1 = df.East
factor3 = df.North
factor2 = df.South

In [19]:
factor_table = np.column_stack((factor1,factor2,factor3))

In [20]:
factor_table

array([[33.53, 40.55, 16.66],
       [36.5 , 36.19, 16.46],
       [34.66, 37.31, 17.66],
       [33.13, 32.52, 17.5 ],
       [35.75, 33.71, 16.4 ],
       [34.46, 34.14, 16.28],
       [34.6 , 34.85, 16.06],
       [35.38, 35.89, 15.93],
       [35.85, 33.53, 16.6 ],
       [35.68, 33.79, 16.41],
       [35.35, 34.72, 16.17],
       [35.04, 35.22, 15.92],
       [34.07, 36.5 , 16.04],
       [32.2 , 37.6 , 16.19],
       [34.32, 37.89, 16.62],
       [31.08, 37.71, 17.37],
       [35.73, 37.  , 18.12],
       [34.11, 36.76, 18.53],
       [34.79, 34.62, 15.54],
       [35.77, 35.4 , 15.7 ],
       [36.44, 35.96, 16.45],
       [37.82, 36.26, 17.62],
       [35.07, 36.34, 18.12],
       [35.26, 35.9 , 19.05],
       [35.56, 31.84, 16.51],
       [35.73, 33.16, 16.02],
       [36.46, 33.83, 15.89],
       [36.26, 34.89, 15.83],
       [37.2 , 36.27, 16.71]])

In [22]:
x = sm.add_constant(factor_table, prepend=True)

In [23]:
x

array([[ 1.  , 33.53, 40.55, 16.66],
       [ 1.  , 36.5 , 36.19, 16.46],
       [ 1.  , 34.66, 37.31, 17.66],
       [ 1.  , 33.13, 32.52, 17.5 ],
       [ 1.  , 35.75, 33.71, 16.4 ],
       [ 1.  , 34.46, 34.14, 16.28],
       [ 1.  , 34.6 , 34.85, 16.06],
       [ 1.  , 35.38, 35.89, 15.93],
       [ 1.  , 35.85, 33.53, 16.6 ],
       [ 1.  , 35.68, 33.79, 16.41],
       [ 1.  , 35.35, 34.72, 16.17],
       [ 1.  , 35.04, 35.22, 15.92],
       [ 1.  , 34.07, 36.5 , 16.04],
       [ 1.  , 32.2 , 37.6 , 16.19],
       [ 1.  , 34.32, 37.89, 16.62],
       [ 1.  , 31.08, 37.71, 17.37],
       [ 1.  , 35.73, 37.  , 18.12],
       [ 1.  , 34.11, 36.76, 18.53],
       [ 1.  , 34.79, 34.62, 15.54],
       [ 1.  , 35.77, 35.4 , 15.7 ],
       [ 1.  , 36.44, 35.96, 16.45],
       [ 1.  , 37.82, 36.26, 17.62],
       [ 1.  , 35.07, 36.34, 18.12],
       [ 1.  , 35.26, 35.9 , 19.05],
       [ 1.  , 35.56, 31.84, 16.51],
       [ 1.  , 35.73, 33.16, 16.02],
       [ 1.  , 36.46, 33.83, 15.89],
 

In [26]:
results = sm.OLS(flux,x).fit()

In [27]:
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               HeatFlux   R-squared:                       0.874
Model:                            OLS   Adj. R-squared:                  0.859
Method:                 Least Squares   F-statistic:                     57.87
Date:                Fri, 02 Jul 2021   Prob (F-statistic):           2.17e-11
Time:                        12:39:42   Log-Likelihood:                -101.39
No. Observations:                  29   AIC:                             210.8
Df Residuals:                      25   BIC:                             216.3
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        389.1659     66.094      5.888      0.000     253.043     525.288
x1             2.1247      1.214      1.750      0.092      -0.376       4.626
x2             5.3185      0.963      5.523      0.000       3.335       7.302
x3           -24.1324      1.869    -12.915      0.000     -27.981     -20.284
==============================================================================
Omnibus:                        1.833   Durbin-Watson:                   1.484
Prob(Omnibus):                  0.400   Jarque-Bera (JB):                0.778
Skew:                           0.328   Prob(JB):                        0.678
Kurtosis:                       3.462   Cond. No.                     2.18e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.18e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

heatflux = 2.1 * East + 5.3 * North - 24.1 * South + 389.16

In [1]:
salary = [34.7,28.9,26.7,32.8,43.6,37.9,28.7]
marks = [67.8,59.3,54.7,65.8,78.6,72.9,58.9]

In [6]:
stats.linregress(marks,salary)

LinregressResult(slope=0.699363979179945, intercept=-12.429814637773546, rvalue=0.9928474346463362, pvalue=8.278321576159311e-06, stderr=0.037610030299421816, intercept_stderr=2.4783774749114253)

y = 0.699 * x - 12.4298

IF THE REGRESSION LINE OFTWO VARIABLES MAKES AN ACUTE ANGLE WITH THE X-AXIS IT INDICATES A POSITIVE CORRELATION AND THE SLOP OFTHE LINE WILL BE POSITIVE

R squared is the correlation coefficient.
The sign of the correlation coefficient shows whether the correlation is positive or negative, a positive sign shows that there's a positive correlation and a negative sign shows that the correlation is negative.
- Positive correlation means the variables are directly proportional
- Negative correlation means the variables are inversely proportional

The closer the mod of correlation is to 1 the stronger the correlation is

i.e as |correlation coefficient| approaches 1 the correlation becomes stronger

standard error = standard deviation / sqrt(n)

it indicates the uncertainity in the prediction of the mean on the basis of the sample data

Ho response does not depend on predictor

In [4]:
df.head()

HeatFlux   East  South  North
0     271.8  33.53  40.55  16.66
1     264.0  36.50  36.19  16.46
2     238.8  34.66  37.31  17.66
3     230.7  33.13  32.52  17.50
4     251.6  35.75  33.71  16.40

In [5]:
resp = list(df.HeatFlux)

In [6]:
east = list(df.East)
south = list(df.South)
north = list(df.North)

In [11]:
x = np.column_stack((east,north,south))
x = sm.add_constant(x, prepend = True)

In [12]:
results = sm.OLS(resp,x).fit()

In [13]:
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.874
Model:                            OLS   Adj. R-squared:                  0.859
Method:                 Least Squares   F-statistic:                     57.87
Date:                Tue, 06 Jul 2021   Prob (F-statistic):           2.17e-11
Time:                        12:51:15   Log-Likelihood:                -101.39
No. Observations:                  29   AIC:                             210.8
Df Residuals:                      25   BIC:                             216.3
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        389.1659     66.094      5.888      0.000     253.043     525.288
x1             2.1247      1.214      1.750      0.092      -0.376       4.626
x2           -24.1324      1.869    -12.915      0.000     -27.981     -20.284
x3             5.3185      0.963      5.523      0.000       3.335       7.302
==============================================================================
Omnibus:                        1.833   Durbin-Watson:                   1.484
Prob(Omnibus):                  0.400   Jarque-Bera (JB):                0.778
Skew:                           0.328   Prob(JB):                        0.678
Kurtosis:                       3.462   Cond. No.                     2.18e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.18e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

There is a 95% chance that the const value will fall between 253.043 and 525.288

As the interval reduces r-square increases, i.e model becomes more accurate

### Equation
y = m1x1 + m2x2 + m3x3 + c

Heatflux = 2.12 * east - 24.13 * north + 5.31 * south + 389.16